Reuse already created FileSystems during collection of ApplicationArchive

Postremus opened the pull request:
The applicationarchive contains paths that are inside a zipfilesystem (of .jar files), needing the filesystem to be kept open during the whole build.
During collection of the applicationarchives by marker files (e.g. META-INF/jandex.idx), a fs is opened for checking for a marker file, and the opened again to be used for the paths of the apparchive.

Also, try to reduce the amount of calls to isDIrectory, since this also showed up in my profiling.

On my (windows) machine, this change saves about 20ms of the dev mode startup time.

Relates to #21552, but not a complete fix.

Postremus commented:
The test failures seem kinda not related. Will do a rebase on main later to pick up latest fixes / let ci re run.

stuartwdouglas commented on core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java:
```diff
@@ -213,24 +215,32 @@ private static void addMarkerFilePaths(Set<String> applicationArchiveFiles,
                     continue;
                 }
-                if (Files.isDirectory(p)) {
+                if (dep.isWorkspacetModule()) {
```
Actually those failures look related, and my guess is that it would be because of this change, not 100% sure why though.

Postremus replied:
The gradle tests fail at `MultiModuleIncludedBuildTest`. This does seem related.

Thank you for the pointer!

Postremus replied:
Done

Postremus committed:
Reuse already created FileSystems during collection of ApplicationArchive
the applicationarchive contains paths that are inside a zipfilesystem (of .jar files), needing the filesystem to be kept open during the whole build.
During collection of the applicationarchives by marker files (e.g. META-INF/jandex.idx), a fs is opened for checking for a marker file, and the opened again to be used for the paths of the apparchive.
Simply reusing the filesystems saves a few cycles. Since, at least on windows, opening new ZipFileSystems is relatively expensive.

Postremus commented:
@stuartwdouglas 
CI is green now, doing another rebase to pickup #21688
